Userland/Libraries/LibJS/Runtime/ObjectEnvironment.cpp

```diff
@@ -127,6 +127,13 @@ ThrowCompletionOr<Value> ObjectEnvironment::get_binding_value(VM&, FlyString con
 {
     auto& vm = this->vm();
 
+    // OPTIMIZATION: In non-strict mode, we don't need the separate HasProperty check since Get
+    //               will return undefined for missing properties anyway. So we take advantage of this
+    //               to avoid doing both HasProperty and Get.
+    // FIXME: We could combine HasProperty and Get in all cases if Get would return a bit more failure information.
+    if (!strict)
+        return m_binding_object.get(name);
```

Comment:
`Proxy` is a thing and ruins the day :grimacing:

```js
const p = new Proxy({}, { has() { console.log("has"); return true; }, get() { console.log("get"); return 42; } });
with (p) { a }
```

FWIW, our current impl has one `has` too many for this case (`has get has get` vs expected `has get get`), but I'm afraid we're skipping the wrong one here :thinking:

Reply:
One (ugly) workaround could be to have a "bool did_override_has_property()" virtual that would be set by all JS::Object descendants that override `internal_has_property`, that we could use here as a condition

Reply:
oh yeah it's not just proxies, pretty sure a bunch of web platform objects do this as well

Reply:
The other option (might be cleaner), is to have this optimisation only when the object environment is the owned by the global object, since I don't think we really care about optimizing `with` accesses, and the global object doesn't override internal_has_property

Reply:
> The other option (might be cleaner), is to have this optimisation only when the object environment is the owned by the global object, since I don't think we really care about optimizing `with` accesses, and the global object doesn't override internal_has_property

This seems like the cleaner option. `ObjectEnvironment` already knows if it's a `with` environment via `[[IsWithEnvironment]]` so we can trivially disable the optimization in that case.